test/envoye2e/driver/fake_stackdriver.go

```diff
@@ -191,9 +204,111 @@ func (s *LoggingServer) GetLogEntries(w http.ResponseWriter, req *http.Request)
 	}
 }
 
+// ListTraces implements ListTraces method.
+func (s *TracesServer) ListTraces(context.Context, *cloudtracev1.ListTracesRequest) (*cloudtracev1.ListTracesResponse, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	return &s.listTracesResp, nil
+}
+
+// GetTrace implements GetTrace method.
+func (s *TracesServer) GetTrace(context.Context, *cloudtracev1.GetTraceRequest) (*cloudtracev1.Trace, error) {
+	log.Fatal("Unimplemented Method")
+	return &cloudtracev1.Trace{}, nil
+}
+
+// GetTrace implements GetTrace method.
+func (s *TracesServer) PatchTraces(context.Context, *cloudtracev1.PatchTracesRequest) (*empty.Empty, error) {
+	log.Fatal("Unimplemented Method")
+	return &empty.Empty{}, nil
+}
+
+func randomHex(n int) (string, error) {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
+
+// BatchWriteSpans implements BatchWriteSpans method.
+func (s *TracesServer) BatchWriteSpans(ctx context.Context, req *cloudtracev2.BatchWriteSpansRequest) (*empty.Empty, error) {
+	log.Printf("receive BatchWriteSpansRequest %+v", *req)
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	var m []*cloudtracev1.Trace
+	// In first Iteration, we find all spans that are root spans and store it in an array.
+	for _, span := range req.Spans {
+		var spanID int
+		var err error
+		if spanID, err = strconv.Atoi(span.SpanId); err != nil {
+			log.Printf("err: %v", err)
+			continue
+		}
+		if span.ParentSpanId == "" {
+			traceID, _ := randomHex(32)
+			m = append(m, &cloudtracev1.Trace{
+				ProjectId: req.Name,
+				TraceId:   traceID,
+				Spans: []*cloudtracev1.TraceSpan{
+					{
+						SpanId:       uint64(spanID),
+						Name:         span.Name,
+						ParentSpanId: 0,
+						StartTime:    span.StartTime,
+						EndTime:      span.EndTime,
+					},
+				},
+			})
+		}
+	}
+	for _, span := range req.Spans {
+		if span.ParentSpanId == "" {
+			continue
+		}
+		var spanID, parentSpanID int
+		var err error
+		if spanID, err = strconv.Atoi(span.SpanId); err != nil {
+			log.Printf("err: %v", err)
+			continue
+		}
+		if parentSpanID, err = strconv.Atoi(span.ParentSpanId); err != nil {
+			log.Printf("err: %v", err)
+			continue
+		}
+
+		for _, trace := range m {
+			for _, parentSpan := range trace.Spans {
+				if parentSpan.SpanId == uint64(parentSpanID) {
```

Comment:
This logic might not work properly, since batch with child span could arrive at server earlier than the parent span, which will cause the span being dropped. So I think you'd have to use trace id inside span name to find out spans belong to the same trace.

Reply:
Replied above.. thanks!